--- init-ifupdown-1.0/init	2017-09-28 22:11:03.707887251 +0000
+++ init-ifupdown-1.0-new/init	2017-09-28 21:59:34.084424122 +0000
@@ -60,6 +60,15 @@
 	echo -n "Configuring network interfaces... "
 	sysctl -e -p /etc/sysctl.conf >/dev/null 2>&1
 	ifup -a
+        ip link set eth0 up
+        mac=$(ifconfig | grep -o "..:..:..:..:..:..")
+        machigh=$(echo $mac | cut -c 13-14 -)
+        maclow=$(echo $mac | cut -c 16-17 -)
+        ip addr add 192.168.$((16#$machigh)).$((16#$maclow))/24 dev eth0
+    #   add default gateway and DNS
+        route add default gw 192.168.1.1 eth0
+        echo "nameserver 8.8.8.8" >> /etc/resolv.conf
+        echo "nameserver 8.8.4.4" >> /etc/resolv.conf
 	echo "done."
 	;;
 
