--- ray-ray-0.2.1/src/common/redis_module/CMakeLists.txt	2017-10-01 19:33:13.000000000 +0000
+++ ray-ray-0.2.1-new/src/common/redis_module/CMakeLists.txt	2017-11-27 08:29:19.636765294 +0000
@@ -15,3 +15,10 @@
 
 target_compile_options(ray_redis_module PUBLIC ${REDIS_MODULE_CFLAGS} -fPIC)
 target_link_libraries(ray_redis_module ${REDIS_MODULE_LDFLAGS})
+install(TARGETS ray_redis_module DESTINATION ${CMAKE_INSTALL_LIBDIR}/ray/core/src/common/redis_module/)
+file(MAKE_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/../../../python/ray/core/src/common/redis_module/)
+add_custom_target(copy_ray_redis_module ALL)
+add_custom_command(TARGET copy_ray_redis_module POST_BUILD
+                   COMMAND ${CMAKE_COMMAND} -E
+                      copy $<TARGET_FILE:ray_redis_module>
+                           ${CMAKE_CURRENT_LIST_DIR}/../../../python/ray/core/src/common/redis_module/)
